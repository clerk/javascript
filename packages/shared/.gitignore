browser
callWithRetry
color
cookie
date
deprecated
error
file
globs
handleValueOrFn
isomorphicAtob
isomorphicBtoa
keys
loadScript
localStorageBroadcastChannel
poller
proxy
underscore
url
react
constants
apiUrlFromPublishableKey