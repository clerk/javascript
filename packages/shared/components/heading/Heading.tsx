import cn from 'classnames';
import React from 'react';

// @ts-ignore
import styles from './Heading.module.scss';

interface HeadingProps extends React.HTMLAttributes<HTMLHeadingElement> {
  as: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
}

export const Heading: React.FC<HeadingProps> = ({
  as,
  children,
  className,
  ...rest
}) => {
  const Heading = as;
  return (
    <Heading {...rest} className={cn(styles.heading, styles[as], className)}>
      {children}
    </Heading>
  );
};
