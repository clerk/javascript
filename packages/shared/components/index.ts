export * from './button';
export * from './card';
export * from './checkbox';
export * from './dropdown';
export * from './form';
export * from './input';
export * from './label';
export * from './modal';
export * from './outline';
export * from './popover';
export * from './portal';
export * from './radio';
export * from './spinner';
export * from './switch';
export * from './tabs';
export * from './titledCard';
export * from './toggle';
export * from './tooltip';
export * from './wizard';
export * from './heading';
export * from './oneTimeCodeInput';
export * from './tag';
export * from './textArea';
export * from './list';
export * from './avatar';
export * from './phoneInput';
export * from './control';
export * from './imageUploader';
