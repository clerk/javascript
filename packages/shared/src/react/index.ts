export * from './hooks';

export {
  ClerkInstanceContext,
  ClientContext,
  OptionsContext,
  OrganizationProvider,
  SessionContext,
  useAssertWrappedByClerkProvider,
  useClerkInstanceContext,
  useClientContext,
  useOptionsContext,
  useOrganizationContext,
  UserContext,
  useSessionContext,
  useUserContext,
} from './contexts';
