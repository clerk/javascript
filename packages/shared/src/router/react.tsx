/**
 * React-specific binding's for interacting with Clerk's router interface.
 */
import React, { createContext, useContext } from 'react';

import type { ClerkHostRouter, ClerkRouter } from './router';
import { createClerkRouter } from './router';

export const ClerkHostRouterContext = createContext<ClerkHostRouter | null>(null);
export const ClerkRouterContext = createContext<ClerkRouter | null>(null);

export function useClerkHostRouter() {
  const ctx = useContext(ClerkHostRouterContext);

  if (!ctx) {
    throw new Error(
      'clerk: Unable to locate ClerkHostRouter, make sure this is rendered within `<ClerkHostRouterContext.Provider>`.',
    );
  }

  return ctx;
}

export function useClerkRouter() {
  const ctx = useContext(ClerkRouterContext);

  if (!ctx) {
    throw new Error('clerk: Unable to locate ClerkRouter, make sure this is rendered within `<Router>`.');
  }

  return ctx;
}

/**
 * Construct a Clerk Router using the provided host router. The router instance is accessible using `useClerkRouter()`.
 */
export function Router({
  basePath,
  children,
  router,
}: {
  children: React.ReactNode;
  basePath?: string;
  router?: ClerkHostRouter;
}) {
  const hostRouter = useClerkHostRouter();
  const clerkRouter = createClerkRouter(router ?? hostRouter, basePath);

  return <ClerkRouterContext.Provider value={clerkRouter}>{children}</ClerkRouterContext.Provider>;
}

type RouteProps = { path?: string; index?: boolean };

/**
 * Used to conditionally render its children based on whether or not the current path matches the provided path.
 */
export function Route({ path, children, index }: RouteProps & { children: React.ReactNode }) {
  const parentRouter = useClerkRouter();

  if (!path && !index) {
    return children;
  }

  if (!parentRouter?.match(path, index)) {
    return null;
  }

  return children;
}
