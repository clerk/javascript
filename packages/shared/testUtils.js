module.exports = require('./utils/testUtils');
