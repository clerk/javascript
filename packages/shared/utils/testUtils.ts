import '@testing-library/jest-dom/extend-expect';

import {render, RenderOptions} from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

import {noop} from '../utils';

// Wrap components with test providers
const customRender = (ui: React.ReactElement, options?: RenderOptions) => {
  return render(ui, options);
};

// To be used for structural snapshot testing
const renderJSON = (c: React.ReactElement) => renderer.create(c).toJSON();

// To be used for structural snapshot testing after the first rerender on mount
const renderJSONAfterFirstAct = (c: React.ReactElement) => {
  let root;

  renderer.act(() => {
    root = renderer.create(c);
  });

  // @ts-ignore
  return root.toJSON();
};

// Re-export everything
export * from '@testing-library/react';
export { act as actHook, renderHook } from '@testing-library/react-hooks';
export { default as userEvent } from '@testing-library/user-event';
export { mocked } from 'jest-mock';

// Override render method
export { customRender as render, renderJSON, renderJSONAfterFirstAct, noop };
