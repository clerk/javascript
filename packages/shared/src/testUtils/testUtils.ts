import '@testing-library/jest-dom/extend-expect';

import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';

import { noop } from '../utils';

// Wrap components with test providers
const customRender = (ui: React.ReactElement, options?: RenderOptions) => {
  return render(ui, options);
};

// Re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Override render method
export { customRender as render, noop };
