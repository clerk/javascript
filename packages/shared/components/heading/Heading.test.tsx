import { renderJSON } from '@clerk/shared/testUtils';
import * as React from 'react';

import { Heading } from './Heading';

describe('<Heading/>', () => {
  it('renders h1 heading', () => {
    const tree = renderJSON(<Heading as="h1">Top heading</Heading>);
    expect(tree).toMatchSnapshot();
  });
});
