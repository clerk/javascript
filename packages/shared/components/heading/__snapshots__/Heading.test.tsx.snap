// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading/> renders h1 heading 1`] = `
<h1
  className="heading h1"
>
  Top heading
</h1>
`;
